### PR TITLE
Add Pnpm AuditAnalyzer

### PR DESCRIPTION
**Pull request by iagotomas**
## Fixes Issue #

feature request: support pnpm and yarn projects #1416

## Description of Change

*Add a new AuditAnalayzer compatible with Pnpm projects*

## Have test cases been added to cover the new functionality?

*yes*

### CODE REVIEW

**Review by iagotomas — COMMENTED** [review]
@jeremylong Could you have a look at the PR? It follows Yarn implementation, hope I didn't miss anything. Will be glad to hear your feedback :)

**Comment by jeremylong** [comment]
Looks good - any chance you would be willing to update the site documentation as well?

https://github.com/jeremylong/DependencyCheck/blob/main/ant/src/site/markdown/configuration.md
https://github.com/jeremylong/DependencyCheck/blob/main/maven/src/site/markdown/configuration.md
https://github.com/jeremylong/DependencyCheck/blob/main/cli/src/site/markdown/arguments.md

I'll get the arguments updated in the gradle plugin with the next release.

**Comment by iagotomas** [comment]
Sure, didn't notice these, thanks for pointing.

**Comment by iagotomas** [comment]
Hi @jeremylong, I fixed the doc as you suggested and conducted some more tests which raised an issue with ProcessReader, not sure if you found it before. The inputstream provided by ProcessReader is topped at 64k and the audit report generated by the npm audit API may exceed that size. To workaround the stdout is redirected to a file and after read to process the full audit report.

File: core/src/main/java/org/owasp/dependencycheck/analyzer/PnpmAuditAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            args.add(getPnpm());
+            args.add("audit");
+            if (skipDevDependencies) {
+                args.add("--prod");
+            }
+            // pnpm audit returns a json compliant with NpmAuditParser
+            args.add("--json");
+            // ensure we are using the right registry despite .npmrc
+            args.add("--registry");
+            args.add("https://registry.npmjs.org/");
```
Comment (jeremylong) [inline core/src/main/java/org/owasp/dependencycheck/analyzer/PnpmAuditAnalyzer.java:211]: Should `registry` be a configurable setting (possibly for npm and yarn also)?

**Review by jeremylong — APPROVED** [review]
Even if the registry should be configurable - we can approve and merge this now. Then add the configuration in a later release.

**Comment by jeremylong** [comment]
It looks like there are test failures:

https://github.com/jeremylong/DependencyCheck/runs/3983551519?check_suite_focus=true#step:5:2124

**Comment by jeremylong** [comment]
Looks like the error is because `pnpm` is not installed on the build system. We likely need to update the workflow to install pnpm.

**Comment by jeremylong** [comment]
Setting up pnpm should be simple with https://github.com/pnpm/action-setup

**Comment by jeremylong** [comment]
I tried to just add the workflow update via #3755 - however, it appears you did not checking one of the test files (`pnpmaudit/pnpm-audit.json`)?

https://github.com/jeremylong/DependencyCheck/pull/3700/files#diff-bf1b785fb925d2d59a7104b2593a7aedbf3bbfcb81c11e58d9259d0ae6a14ce8R30

File: core/src/main/java/org/owasp/dependencycheck/analyzer/PnpmAuditAnalyzer.java (hunk 1 above)
Comment (iagotomas) [reply core/src/main/java/org/owasp/dependencycheck/analyzer/PnpmAuditAnalyzer.java:211]: Think would be great to have a common registry setting for all, in our company we use them all and plan to use owasp, we only missed pnpm :)

**Comment by iagotomas** [comment]
Sorry missed the file, just added, hope it works now

**Comment by jeremylong** [comment]
Merged via #3755

**Pull request closed without merging**